### PR TITLE
Encoder refactor V2

### PR DESCRIPTION
**Pull request by dantreiman**
This PR contains all of Connor's commits from https://github.com/ludwig-ai/ludwig/pull/2269 (`:encoder_refactor`) merged atop the master branch, plus additional refactoring and test fixes in `:daniel/encoder_refactor`
```
master -----------------+------+>
                        |      |
daniel/encoder_refactor |      +*---->
                        |       | 
encoder_refactor        +------->

+ = branch
* = merge commit
```

A few notes:
- We have a glue function `load_config_with_kwargs` which loads a config from a dictionary, and collects any key-value pairs not captured by the config in an extra `kwargs` dictionary.  I removed this from instantiation of input features, now all input feature and encoder params are captured by the config.
- `load_config_with_kwargs` is still in use for combiners to pass in `output_size`.

### CODE REVIEW

File: ludwig/constants.py
Comment (justinxzhao) [inline ludwig/constants.py:66]: nit: This is duplicated.

File: ludwig/features/audio_feature.py
Comment (justinxzhao) [inline ludwig/features/audio_feature.py:476]: nit: This could be one statement.

```python
set_default_values(
    input_feature, 
    {
        TIED: defaults.tied, 
        PREPROCESSING: {}, 
        ENCODER: {TYPE: defaults.encoder.type}
    }
)
```

File: ludwig/features/base_feature.py
Comment (justinxzhao) [inline ludwig/features/base_feature.py:125]: Nice! Glad you were able to make this change.

File: ludwig/features/category_feature.py
Comment (justinxzhao) [inline ludwig/features/category_feature.py:162]: nit: Remove.

**Review by justinxzhao — APPROVED** [review]
Minor nits. LGTM.

File: ludwig/features/base_feature.py
Comment (connor-mccorm) [inline ludwig/features/base_feature.py:198]: Just a quick question, but why do we load the config in the specific input/output features and the base output feature, but not in the base input feature?

File: ludwig/constants.py
Comment (dantreiman) [reply ludwig/constants.py:66]: Pre-commit merge error, good catch!

**Review by connor-mccorm — COMMENTED** [review]
Thank you so much for finishing this off. Just had a small question about `load_config` in features/base.py, and also wanted to discuss a piece of feedback that Piero gave in the older PR but besides that I think it should be good to merge. Happy to chat now about these points or we can talk in standup at 11.

File: ludwig/features/base_feature.py
Comment (dantreiman) [reply ludwig/features/base_feature.py:198]: This is vestigial code - My first instinct was to `load_config` in the base class, but then I changed my mind.  `load_config` in `OutputFeature` doesn't do anything right now, will remove.

File: ludwig/data/dataset_synthesizer.py
Comment (dantreiman) [reply ludwig/data/dataset_synthesizer.py:116]: I recommend doing that as a follow-up task after this PR is merged, because what we have here works and the follow-up task is parallelizable AFAIK nothing else we're doing depends on it.  Created a follow-up task in linear here:  https://linear.app/predibase/issue/MLQ-54/refactor-usage-of-encoder-configdecoder-config-features

**Comment by dantreiman** [comment]
I recommend doing https://github.com/ludwig-ai/ludwig/pull/2370#pullrequestreview-1072904998 as a follow-up task after this PR is merged, because what we have here works and the follow-up task is parallelizable AFAIK nothing else we're doing depends on it.  Created a follow-up task in linear: here:  https://linear.app/predibase/issue/MLQ-54/refactor-usage-of-encoder-configdecoder-config-features

**Review by connor-mccorm — APPROVED** [review]
LGTM

File: ludwig/features/category_feature.py
Comment (tgaddair) [inline ludwig/features/category_feature.py:420]: Is `populate_defaults()` needed anymore? I believe all defaults should be encoded in the schema, so I don't think we need to set them explicitly, but perhaps I'm missing something.

File: ludwig/features/date_feature.py
Comment (tgaddair) [inline ludwig/features/date_feature.py:118]: Is it necessary to call `load_config` here? If the `input_feature_config` is always a `DateInputFeatureConfig`, then it should be a no-op, right?

File: ludwig/features/feature_utils.py
Comment (tgaddair) [inline ludwig/features/feature_utils.py:119]: I thought the decoder was schema-fied as well. Is that not the case?

File: ludwig/features/h3_feature.py
Comment (tgaddair) [inline ludwig/features/h3_feature.py:114]: Remove commented code.

**Review by tgaddair — APPROVED** [review]
Amazing work! I had a couple comments, but no worries if the answer is "we'll do this in the follow-up". But I would like to know if we're on the same page about the ideal state, and what the blockers are to getting there.

File: ludwig/features/feature_utils.py
Comment (connor-mccorm) [reply ludwig/features/feature_utils.py:119]: Ya you're right, this is a good catch. I think the dependencies aren't actually being acknowledged here which is why this isn't throwing an error. Looking into this.

File: ludwig/features/category_feature.py
Comment (connor-mccorm) [reply ludwig/features/category_feature.py:420]: So I believe there are a couple of defaults that still need to be set with `populate_defaults()` before it gets to the actual feature initialization, IIRC for the preprocessing step. I will double check on this. Either way we will be removing this once the full config object lands so you are correct that they shouldn't be needed anymore with the new schemas, just think theres a small amount of additional wiring needed to fully remove this.

File: ludwig/features/date_feature.py
Comment (connor-mccorm) [reply ludwig/features/date_feature.py:118]: So I believe @dantreiman implemented this section so I will defer to him for the architecture choice here. What I can say is that we should be able to remove this once the config object PR is landed since the full config object will be initialized earlier in the pipeline so we won't have to load the individual configs here in the feature initializers.

Comment (dantreiman) [reply ludwig/features/date_feature.py:118]: Woops.  I meant to declare the type of `input_feature_config` as `Union[dict, DateInputFeatureConfig]`.  I intended to keep that initialization path working until we finally finish schema-fication because many of our tests construct input features from a config dict.

File: ludwig/features/feature_utils.py
Comment (dantreiman) [reply ludwig/features/feature_utils.py:119]: Wow thanks for catching that: this code would have failed if it had ever been called.  I'll update it and add a new test to `test_dependencies.py`

File: ludwig/features/category_feature.py
Comment (dantreiman) [reply ludwig/features/category_feature.py:423]: For now, `populate_defaults()` is needed for preprocessing, since we don't instantiate the feature objects until after preprocessing.  We don't need an implementation per class anymore though, because all of the defaults can be stored in the schema.  @connor-mccorm is planning to remove populate_defaults in his config object PR, which will be a follow-up to this one.

File: ludwig/schema/features/base.py
Comment (connor-mccorm) [inline ludwig/schema/features/base.py:29]: Shouldn't allow none for `name` and `type` be `false`?

File: ludwig/features/bag_feature.py
Comment (w4nderlust) [inline ludwig/features/bag_feature.py:98]: I'm not a fan of setting the encoder_config to self and then inside self.initialize_encoder() referring to it through self. I would rather pass the encoder_config explicitly to the initialize_encoder function.

Comment (w4nderlust) [reply ludwig/features/bag_feature.py:98]: This applies to basically every input feature type

Comment (connor-mccorm) [reply ludwig/features/bag_feature.py:98]: I'm actually working on this as we speak. Changing this to a method where we pass the encoder config into the encoder initializer to create the encoder. The encoder config will then live on the encoder object. This will avoid any issues with tied weights pointing to one encoder object but an encoder config pointing to another.

File: ludwig/features/feature_utils.py
Comment (w4nderlust) [inline ludwig/features/feature_utils.py:120]: previously the fc stack of each output feature was not pert of the decoder, was like a inbetween step between the combiner output and the decoder input. For that reason, parameters like num_fc_stack where assigned to the output feature and not to the decoder. Not sure if this was changed, but here I see something weird: the num_fc_layers is a parameter of the decoder, but in the following line the fc_stack from which the output shape is obtained is not inside the decoder but it's a property of the output feature (like it was before), so maybe this needs to be reconsiliated.

File: ludwig/features/bag_feature.py
Comment (w4nderlust) [inline ludwig/features/bag_feature.py:94]: just a note to remind of the inconsistency of the self.encoder_config assignment in case there is a tied encoder and so encoder_obj is not None. This applies to all input feature types

File: ludwig/schema/features/base.py
Comment (dantreiman) [reply ludwig/schema/features/base.py:29]: If allow_none is false, then we must provide a (non-None) default, so I left allow_none True.  What would be the default if allow_none False?

File: ludwig/features/bag_feature.py
Comment (connor-mccorm) [reply ludwig/features/bag_feature.py:94]: Great catch, I'm actually working on changing this right now which will address this comment and your first comment!

File: ludwig/features/sequence_feature.py
Comment (w4nderlust) [inline ludwig/features/sequence_feature.py:270]: we can revise this later, but I'm not a fan of having two parameters for vocab and vocab size around

File: ludwig/features/feature_utils.py
Comment (dantreiman) [reply ludwig/features/feature_utils.py:120]: Output features still have an fc_stack pre-decoder.  This is my mistake, will fix

Comment (dantreiman) [reply ludwig/features/feature_utils.py:120]: Whoa, but we did move these params to the decoder though we still instantiate the FC stack in OutputFeatureBase:

File: ludwig/trainers/trainer_lightgbm.py
Comment (w4nderlust) [inline ludwig/trainers/trainer_lightgbm.py:482]: I believe num_classes should likely be a property of the output feature, not the decoder, as here for the lightgbm trainer it should not be required to provide a decoder config to begin with. @jppgks to doublecheck

File: ludwig/features/feature_utils.py
Comment (w4nderlust) [reply ludwig/features/feature_utils.py:120]: I think either is fine, if those parameters end up being decoder ones or output feature ones, we should be clear though about which is which and to have consistency ;)

File: ludwig/schema/features/base.py
Comment (connor-mccorm) [reply ludwig/schema/features/base.py:29]: This is a good point. I realized that since the json schema mapping still has them as required params, this won't cause an issue with users potentially not specifying a name or feature type so we can disregard this!

File: ludwig/features/bag_feature.py
Comment (connor-mccorm) [reply ludwig/features/bag_feature.py]: Just finished this switch so we no longer use self in the initialization of encoders and decoders. The config is directly piped in and then becomes an attribute of the `encoder_obj` / `decoder_obj`

File: ludwig/features/bag_feature.py
Comment (connor-mccorm) [reply ludwig/features/bag_feature.py]: This should be resolved now that the encoder and decoder configs live on the `encoder_obj` / `decoder_obj` now.

File: ludwig/features/sequence_feature.py
Comment (connor-mccorm) [reply ludwig/features/sequence_feature.py]: Agreed it is a little clunky and should probably be set more elegantly. I will create an issue on my Linear board to refactor this in a followup PR.

File: ludwig/trainers/trainer_lightgbm.py
Comment (connor-mccorm) [reply ludwig/trainers/trainer_lightgbm.py:482]: Awaiting @jppgks reply to determine next steps!

File: ludwig/features/feature_utils.py
Comment (dantreiman) [reply ludwig/features/feature_utils.py]: After investigating this I made some additional changes:

- Moved fc_stack parameters from decoders back into output feature
- Made threshold a property of BinaryOutputFeatureConfig, not the decoder
- Similarly made clip a property of NumberOutputFeatureConfig
- Output feature's `input_size` should be the output of the combiner (plus any dependencies).  Decoder's `input_size` is the output size of the fc stack.

File: ludwig/trainers/trainer_lightgbm.py
Comment (jppgks) [reply ludwig/trainers/trainer_lightgbm.py]: Agree with @w4nderlust

Comment (jppgks) [reply ludwig/trainers/trainer_lightgbm.py]: by default, GBM uses the passthrough decoder https://github.com/ludwig-ai/ludwig/blob/4405ef8745aab886e973dcdf3ad77c2048bf9f74/ludwig/utils/defaults.py#L313

**Pull request merged**
